lib/matchSourceExpression.js

```diff
@@ -0,0 +1,180 @@
+var urlModule = require('url');
+
+var ipv4DigitRegExp = /([0-9]|1?[0-9][0-9]|2[0-4][0-9]|25[0-5])/;
+var ipv4RegExp = new RegExp(
+    '^' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '\\.' +
+    ipv4DigitRegExp.source + '$'
+);
+
+var defaultPortByScheme = {
+    http: 80,
+    https: 443
+};
+
+function origin(url) {
+    if (/file:\/\//.test(url)) {
+        return 'file:';
+    }
+    var matchOrigin = url.match(/^([a-z][a-z0-9+.-]+):\/\/(\*|(?:\*\.)?[a-z0-9-]+(?:\.[a-z0-9-]+)*)(?::(\d+|\*))?(\/.*)?$/i);
+    if (matchOrigin) {
+        var scheme = matchOrigin[1];
+        var host = matchOrigin[2];
+        var port = matchOrigin[3] || defaultPortByScheme[scheme];
+        return scheme + '://' + host + ':' + port;
+    }
+}
+
+function slashSplit(str) {
+    var tokens = str.replace(/^\/|\/$/g, '').split('/');
+    if (tokens.length === 1 && tokens[0] === '') {
+        return [];
+    } else {
+        return tokens;
+    }
+}
+
+// https://www.w3.org/TR/CSP/#match-source-expression
+module.exports = function matchSourceExpression(url, sourceExpression, protectedResourceUrl) {
+    var protectedResourceUrlObj = urlModule.parse(protectedResourceUrl);
+    // Let url be the result of processing the URL through the URL parser.
+    var urlObj = urlModule.parse(url);
+
+    // If the source expression a consists of a single U+002A ASTERISK character (*), and url’s scheme is not one of blob, data, filesystem, then return does match.
+    if (sourceExpression === '*' && urlObj.protocol !== 'data:' && urlObj.protocol !== 'blob:' && urlObj.protocol !== 'filesystem:') {
+        return true;
+    }
+
+    // If the source expression matches the grammar for scheme-source (https://www.w3.org/TR/CSP/#scheme_source):
+    if (/^[a-z][a-z0-9+.-]*:$/i.test(sourceExpression)) {
+        // If url’s scheme is an ASCII case-insensitive match for the source expression’s scheme-part, return does match.
+        // Otherwise, return does not match.
+        return urlObj.protocol.toLowerCase() === sourceExpression.toLowerCase();
+    }
+
+    // If the source expression matches the grammar for host-source (https://www.w3.org/TR/CSP/#host_source)
+    var matchHostSource = sourceExpression.match(/^(?:([a-z][a-z0-9+.-]+):\/\/)?(\*|(?:\*\.)?[a-z0-9-]+(?:\.[a-z0-9-]+)*)(?::(\d+|\*))?(\/.*)?$/i);
```

Comment:
This regex only has one optional capture group difference to `matchOrigin`, should `matchOrigin` be reused?